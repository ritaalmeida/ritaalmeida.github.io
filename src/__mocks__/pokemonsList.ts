export const pokemons = [
  {
    id: 4,
    name: 'charmander',
  },
  {
    id: 5,
    name: 'charmeleon',
  },
  {
    id: 6,
    name: 'charizard',
  },
  {
    id: 390,
    name: 'chimchar',
  },
  {
    id: 737,
    name: 'charjabug',
  },
  {
    id: 935,
    name: 'charcadet',
  },
  {
    id: 1025,
    name: 'pecharunt',
  },
];

export default pokemons;
