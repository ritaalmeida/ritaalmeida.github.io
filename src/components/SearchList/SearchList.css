.searchInput {
  background-image: url('../../assets/searchicon.png');
  background-size: 18px;
  background-position: 16px 12px;
  background-repeat: no-repeat;
  width: 100%;
  font-size: 16px;
  padding: 12px 20px 12px 46px;
  border: 1px solid #ddd;
  border-radius: 24px;
}

.list {
  list-style-type: none;
  padding: 0 14rem;
  margin: 0;
  width: 100%;
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 50%;
  left: 50%;
  overflow: scroll;
  max-height: 325px;
  transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
}

.list li a {
  border-top: 1px solid #ddd;
  background-color: #f6f6f6;
  padding: 12px;
  text-decoration: none;
  text-transform: capitalize;
  font-size: 18px;
  color: black;
  display: block;
  animation: fadeIn 0.5s ease-in-out;

}

.list li a:hover {
  background-color: #eee;
}

.pokemonItem{
  display: flex;
  justify-content: space-between;
}

.pokemonInfo {
  font-size: small;
  color: gray;
}

@keyframes fadeIn {
  from {
    opacity: 0;
    transform: translateY(-45px);
  }

  to {
    opacity: 1;
    transform: translateY(0);
  }
}
