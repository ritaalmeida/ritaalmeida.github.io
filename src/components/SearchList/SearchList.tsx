import {
  ChangeEvent,
  Dispatch,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import "./SearchList.css";
import { Pokemon } from "../../interfaces";

export interface SearchListProps {
  searchText: string;
  setSearchText: Dispatch<SetStateAction<string>>;
  pokemons?: Pokemon[];
  handleSelectPokemon: (id: number) => void;
}

export const SearchList = ({
  searchText,
  setSearchText,
  pokemons,
  handleSelectPokemon,
}: SearchListProps) => {
  const [, setDebouncedInputValue] = useState("");

  useEffect(() => {
    const delayInputTimeoutId = setTimeout(() => {
      setDebouncedInputValue(searchText);
    }, 500);
    return () => clearTimeout(delayInputTimeoutId);
  }, [searchText]);

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchText(e.target.value);
  };

  return (
    <div className="searchList">
      <input
        type="text"
        className="searchInput"
        placeholder="Search for pokemons"
        title="Type in a name"
        onChange={handleInputChange}
        value={searchText}
      />
      {pokemons && (
        <ul className="list">
          {pokemons?.map((pokemon: Pokemon) => (
            <li key={`pokemon-${pokemon.id}`}>
              <div
                role="menuitem"
                className="menuItem"
                onClick={() => handleSelectPokemon(pokemon.id)}
              >
                <div className="pokemonItem">
                  <span>{pokemon.name}</span>
                  <span className="pokemonItemInfo">{`#${pokemon.id}`}</span>
                </div>
              </div>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};
