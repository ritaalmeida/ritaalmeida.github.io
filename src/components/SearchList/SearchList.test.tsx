import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SearchList, SearchListProps } from './SearchList';
import { pokemons } from '../../__mocks__/pokemonsList';

describe('SearchList', () => {
  let props: SearchListProps;

  beforeEach(() => {
    props = {
      searchText: '',
      setSearchText: jest.fn(),
      handleSelectPokemon: jest.fn(),
    };
  });

  it('should render the component', () => {
    const { container } = render(<SearchList {...props} />);
    expect(container).toMatchSnapshot();
  });

  describe('when the input value changes', () => {
    it('should call setSearchText', () => {
      render(<SearchList {...props} />);

      userEvent.type(screen.getByRole('textbox'), 'test');

      expect(props.setSearchText).toHaveBeenCalled();
    });
  });

  describe('when there are options to list', () => {
    beforeEach(() => {
      props = {
        ...props,
        pokemons,
      };
    });

    it('should render with list', () => {
      const { container } = render(<SearchList {...props} />);
      expect(container).toMatchSnapshot();
    });

    describe('when the user clicks on a option', () => {
      it('should call handleSelectPokemon', () => {
        render(<SearchList {...props} />);

        userEvent.click(screen.getAllByRole('menuitem')[0]);

        expect(props.handleSelectPokemon).toHaveBeenCalledWith(pokemons[0].id);
      });
    });
  });
});
