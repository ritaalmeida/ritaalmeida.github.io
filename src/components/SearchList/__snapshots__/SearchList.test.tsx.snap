// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchList should render the component 1`] = `
<div>
  <div
    class="searchList"
  >
    <input
      class="searchInput"
      placeholder="Search for pokemons"
      title="Type in a name"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`SearchList when there are options to list should render with list 1`] = `
<div>
  <div
    class="searchList"
  >
    <input
      class="searchInput"
      placeholder="Search for pokemons"
      title="Type in a name"
      type="text"
      value=""
    />
    <ul
      class="list"
    >
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            charmander
          </span>
          <span
            class="pokemonItemInfo"
          >
            #4
          </span>
        </div>
      </li>
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            charmeleon
          </span>
          <span
            class="pokemonItemInfo"
          >
            #5
          </span>
        </div>
      </li>
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            charizard
          </span>
          <span
            class="pokemonItemInfo"
          >
            #6
          </span>
        </div>
      </li>
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            chimchar
          </span>
          <span
            class="pokemonItemInfo"
          >
            #390
          </span>
        </div>
      </li>
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            charjabug
          </span>
          <span
            class="pokemonItemInfo"
          >
            #737
          </span>
        </div>
      </li>
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            charcadet
          </span>
          <span
            class="pokemonItemInfo"
          >
            #935
          </span>
        </div>
      </li>
      <li
        class="menuItem"
        role="menuitem"
      >
        <div
          class="pokemonItem"
        >
          <span>
            pecharunt
          </span>
          <span
            class="pokemonItemInfo"
          >
            #1025
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
